Add --force-color flag

CodingItWrong opened the pull request:
Fixes #2230 by adding a --force-color flag to use alongside --no-color, while preserving backward-compatibility with the --color flag.

Please let me know anything I can do to improve this PR! A few thoughts:

- configuration.rb is a bit more complex than it otherwise might need to be because of --color backward compatibility, but I think it's worth it
- Setting config.no_color = true in spec_helper.rb felt a bit dirty, but I do think the best default behavior in the specs is to _not_ have color in the output, and that was the most reliable way I could find to accomplish that.
- The specs that distinguish the behavior of force_color and no_color when --tty is passed can probably be removed, now that they've demonstrated that that flag doesn't affect their behavior.
- The name of the --force-color flag gets a bit confusing with the ConfigurationOptions method `force` (e.g. it "forces force_color"), but I think --force-color is the best external name, and it probably doesn't make sense to consider any renaming of the internal `force` method just for this.
- Reading `no_color = true` and `no_color = false` gets a bit confusing, but this wouldn't extend to end users: they just use --force-color or --no-color and get the expected result. The two different properties are needed, as I describe in the commit message for 9af8f0c, because the default behavior is different from both the force_color and no_color behavior, so there are several different states.

myronmarston commented:
Thanks for putting this thorough PR together!  I'm going to try to review it later in the weekend.

myronmarston commented on lib/rspec/core/configuration.rb:
```diff
@@ -778,7 +780,7 @@ def full_backtrace=(true_or_false)
       end
 
-      # Returns the configuration option for color, but should not
+      # Deprecated: returns the configuration option for color, but should not
```
YARD has built in support for a `@deprecated` tag. Can you use that instead?  Example:

https://github.com/rspec/rspec-core/blob/6a4376e276e52b89d35aa5e16b74758e9f64b991/lib/rspec/core/example.rb#L108

http://rspec.info/documentation/3.5/rspec-core/RSpec/Core/Example.html#rerun_argument-instance_method

myronmarston replied:
Also, it'd be nice if the docs explained why it was deprecated and what the user's options are (e.g. set nothing or `--force-color` instead).

myronmarston commented on lib/rspec/core/configuration.rb:
```diff
@@ -787,17 +789,49 @@ def color
       end
 
-      # Toggle output color.
+      # Deprecated: toggle output color.
```
Same here -- please use the YARD `@deprecated` tag.

myronmarston replied:
Can you deprecate `--tty` as well?  I believe it's unneeded now that `--force-color` is a thing.

CodingItWrong replied:
That definitely makes sense. The only reference to `tty` I see in this file is the `add_setting :tty` on line 401. That's already marked as `@private` and I don't see it generated in the docs. Thinking I should mark it both `@private` and `@deprecated`? Or make it no longer `@private` so as to say "hey, this exists as you well know, but stop using it`? ;-)

myronmarston replied:
Let's keep it private, but add a `@deprecated` tag on it, anyway.  While it's private, it's something that has never been officially public, there are external uses of it, and treating it as a deprecated (including adding a warning in 3.99 and removal in 4.0) won't be too hard here, so the `@deprecated` tag will make it's status clear.

myronmarston commented on lib/rspec/core/option_parser.rb:
```diff
@@ -140,8 +140,18 @@ def parser(options)
-          options[:color] = o
+        parser.on('-c', '--color', '--colour', 'Deprecated: enable color in the output.') do |_o|
+          options[:color] = true
+        end
```
I think maybe we should not list `--color` in `--help` since it is deprecated (but still support it).  We don't really want to take up valuable `--help` real estate with something deprecated.

OTOH, it is still supported, so I can see an argument for keeping it in `--help`.  Thoughts?  (Especially from @rspec/rspec people...)

CodingItWrong replied:
(Note to self: I'll defer acting on this change until more input from RSpec team. I don't have an opinion either way.)

myronmarston replied:
The more I think about it, the more I lean towards removing it.  We need to not break when people pass `-c` or `--color` for backwards compatibility but we don't really want to waste new user's time having them learn about it when they run `rspec --help`.

In the absence of anyone arguing to keep it in the help, I'm going to say let's remove it.

myronmarston commented on lib/rspec/core/option_parser.rb:
```diff
@@ -140,8 +140,18 @@ def parser(options)
+        parser.on('--no-color', '--no-colour', 'Force the output to not be in color.') do |_o|
+          options[:no_color] = true
+          options.delete(:force_color)
         end
```
It'd be nice to find a way to list these two options as one (particularly if we are dropping `--color` from `--help`) similar to the `--[no-]color` thing we have now -- although I'm not sure what the standard is for CLI tools when you are trying to show `--force-color` or `--no-color`.  If there's no good way to combine them in the output we can keep them separate.

Regardless, it'd be nice if the description mentioned something like:

* `Force color on or off, ignoring TTY status of output` (e.g. if we combine them)
* `Force the output to be in color, even if the output is not a TTY` (for `--force-color`)
* `Force the output to not be in color, even if the output is a TTY` (for `--no-color`)

CodingItWrong replied:
I'm having trouble finding documentation on how the `OptionParser` sees that a flag has an optional `[no]` and returns a `false` instead of a `true`. But it seemed like I might be able to detect `'--[force][no]-color'` instead, in order to treat this as a single option for both detection and help output. But when I tried that, neither `--force-color` nor `--no-color` was detected—so I think I'm out of my depth finding a way to combine them. Can anyone direct me to something else I can investigate to try?

myronmarston replied:
Let's just go with the two separate flags like you have it.

myronmarston commented on lib/rspec/core/option_parser.rb:
```diff
@@ -140,8 +140,18 @@ def parser(options)
+
+        parser.on('--force-color', '--force-colour', 'Force the output to be in color.') do |_o|
+          options[:force_color] = true
+          options.delete(:no_color)
```
Why is it necessary to delete the other option here and below?

CodingItWrong replied:
Previously, we used the built-in `OptionParser` behavior where passing `--no-color` would just unset the `:color` option, since there were only two states: `:color` set, or default. But now there are three needed states:

1. Default behavior: show color iff tty
2. Show color regardless of tty
3. Do not show color, regardless of tty

So we need a way to indicate any of those states to the configuration. I opted for having a `:force_color` option and `:no_color` option, so that they directly correspond to the two options.

So then the question is, what to do if both are set? The previous behavior of the system was that whichever flag was set last wins. But now that they are two different options, setting both flags would just set both options to true. And there would be no way for the configuration to tell the order they were set in: it just receives the two booleans.

So, to preserve the behavior of the last-flag-winning, whenever `:force_color` or `:no_color` is set, I unset the opposite flag.

Thoughts?

myronmarston replied:
It sounds like this would make the first flag win instead of last flag wins -- because when the first of the two flags is processed, it deletes the other one, and the last flag is therefore never processed.

One alternate idea is to abort with a message telling the user to set only one of `--no-color` and `--force-color`.  I think I like that better -- instead of guessing their intent, make them correct the command.  That could be done by adding a validation to the `parse` method (or to a new `validate` helper method called from `parse`) that `abort`s with a useful message if both `option[:no_color]` and `options[:force_color]` are `true`.

CodingItWrong replied:
https://github.com/rspec/rspec-core/pull/2321/files#diff-2bae3d9145ad46203f8a7fa5140ce8acL200 actually demonstrates that the last flag wins. The way this works is because the flags don't delete each other. Instead, the `parser` processes each flag in turn, which sets and unsets _something else_: an entry in the `options` hash. That's why unsetting an option doesn't delete a flag.

I can definitely look into the aborting option. Question for backward compatibility: the `--color` and `--no-color` flags previously had this last-wins behavior, so I think it would be good to leave that in. So I'm thinking just abort in case `--force-color` and `--no-color` are used together. Sound good?

myronmarston replied:
Thanks for the explanation, and sounds good!

CodingItWrong replied:
OK, this is in. I actually check at the `configuration.rb` options level instead of at the flag-parsing level, so that the issue will be caught whether the user is using command-line flags or setting the configs directly. However, I'm running into an issue where `abort` is being called twice:

```
~/apps/battle_logic [master] # bin/rspec --force-color --no-color
Please use only one of --force-color and --no-color

Please use only one of --force-color and --no-color
```

I hate to ask you to do my debugging for me, but could you check out where that's implemented and see if a cause or a better place to put that `abort` jumps out at you? Here it is:

https://github.com/rspec/rspec-core/pull/2321/commits/e8a31a6ebfdcc82ee1af13f5b0e8a345c6f8aa25

myronmarston commented on spec/rspec/core/configuration_options_spec.rb:
```diff
@@ -198,16 +210,24 @@
+
+    it "overrides previous :force_color => true" do
+      expect(parse_options('--force-color', '--no-color')).to include(:no_color => true)
+      expect(parse_options('--force-color', '--no-color')).to_not include(:force_color => true)
```
This can be combined into one compound expectation:

``` ruby
expect(parse_options('--force-color', '--no-color')).to include(:no_color => true).and exclude(:force_color => true)
```

myronmarston commented on spec/rspec/core/configuration_options_spec.rb:
```diff
@@ -198,16 +210,24 @@
-      expect(parse_options('--no-color', '--color')).to include(:color => true)
+    it "overrides previous :no_color => true" do
+      expect(parse_options('--no-color', '--force-color')).to include(:force_color => true)
+      expect(parse_options('--no-color', '--force-color')).to_not include(:no_color => true)
```
Same here.

myronmarston commented on spec/rspec/core/configuration_spec.rb:
```diff
@@ -1229,7 +1229,171 @@ def metadata_hash(*args)
+    describe "color options" do
+      before { config.no_color = false }
+
+      context "#force_color=false, #no_color=false" do
```
This assumes that these options default to false.  I think it'd be better to express that assumption with some expectations in a `before` hook (or set them explicitly and do not assume it at all).

myronmarston commented on spec/rspec/core/configuration_spec.rb:
```diff
@@ -1229,7 +1229,171 @@ def metadata_hash(*args)
+            output = StringIO.new
+            config.output_stream = output
+
+            allow(config.output_stream).to receive_messages :tty? => true
```
Not a merge blocker, but every where else our style for this kind of thing is to use parens for the `recieve_messages` call:

``` ruby
allow(config.output_stream).to receive_messages(:tty? => true)
```

It'd be nice for consistency reasons if you could update this.

myronmarston commented on spec/rspec/core/configuration_spec.rb:
```diff
@@ -1229,7 +1229,171 @@ def metadata_hash(*args)
+
+            allow(config.output_stream).to receive_messages :tty? => true
+
+            expect(config.color_enabled?).to be_truthy
```
I'd prefer `be true` over `be_truthy` and `be false` over `be_falsey` -- mind updating the expectations?

myronmarston commented on spec/rspec/core/configuration_spec.rb:
```diff
@@ -1229,20 +1229,124 @@ def metadata_hash(*args)
-            expect(config.color_enabled?).to be_falsy
-            expect(config.color_enabled?(output)).to be_falsy
+            expect(config.color_enabled?).to be_truthy
+            expect(config.color_enabled?(output)).to be_truthy
```
On the surface, it seems a little unfortunate to change the behavior here.  In theory, it would be more backwards compatible to make explicit calls to `config.color = false` behave the same as before and only do the deafulting described here if `config.color` is unset (e.g. `nil`).  In practice, I'm not sure that it's worth the effort.

Thoughts?

CodingItWrong replied:
I think that depends on how often `config.color` is set in practical usage. If I understand how the CLI works, then there isn't actually a way to set `config.color = false`. There wasn't before, either: previously `:color` was either set to `true` or not set, and now at most one of `:force_color` and `:no_color` is set to true, and also `:color` is either set to `true` or not.

Can you think of a real-usage case where `config.color` would be set to `false`?

(Also, this makes me realize that I think `--no-color` should unset not only `:force_color`, but also `:color`. This would preserve the behavior if someone used the flags `--color --no-color`. I'll make a commit with that change.)

myronmarston replied:
> Can you think of a real-usage case where `config.color` would be set to false?

Here's a real-world case I found by searching github:

https://github.com/joecorcoran/judge/blob/971df0a90f9e048b6f9f54c340450bf122e27f8b/spec/spec_helper.rb#L11

CodingItWrong replied:
Gotcha, didn't think about setting the config that way.

An even-more-esoteric case: if someone manually sets `config.color = nil`, in the past that would have behaved the same as `config.color = false`, but if we make this change then the two cases will be have differently. In reality I don't see any instances of that config setting on GitHub. But it does get into some of the complexity this change would entail.

You said " In practice, I'm not sure that it's worth the effort."—so I think you'd be the best to make the call! Code to distinguish the `false` case that exists in the wild from the `nil` case that doesn't?

myronmarston commented on spec/rspec/core/source/syntax_highlighter_spec.rb:
```diff
@@ -2,7 +2,7 @@
 class RSpec::Core::Source
   RSpec.describe SyntaxHighlighter do
-    let(:config)      { RSpec::Core::Configuration.new.tap { |c| c.color = true } }
+    let(:config)      { RSpec::Core::Configuration.new.tap { |config| config.no_color = false } }
```
Does `no_color = false` actually do anything?  Is it necessary to modify the config here at all?

CodingItWrong replied:
Yes, it overrides `spec_helper.rb:13` where I set `config.no_color = true`. This setting preserves the previous assumption that the suite had that color was disabled by default. And I think that's the right default to continue to have:  expectations against plain text are more readable. For example, notifications_spec.rb:232 expects "1.1) Failure/Error", not "1.1) \e[31mFailure/Error"—which is what it gets when color is enabled. Also, I'm not 100% sure but I would assume there would be a way to run specs in or out of a TTY, and we don't want the specs' color settings to vary based on that.

This file needs color enabled, so, since `config.no_color = true` is set in the `spec_helper`, I need to unset it here. Otherwise, for example, the expectation to `be_highlighted` on line 14 fails.

myronmarston replied:
Thanks for the explanation.  I mis-understood how you were setting `no_color = true` in `spec_helper.rb` -- I didn't realize you were monkey patching (or updating and existing monkey patch) to make _all_ instances of `RSpec::Core::Configuration` in our spec suite default to `no_color = true`.

Instead, can you move `config.no_color = true` to the `sandboxed` block in [spec/support/sandboxing.rb](https://github.com/rspec/rspec-core/blob/6a4376e276e52b89d35aa5e16b74758e9f64b991/spec/support/sandboxing.rb#L7)?  The sandboxing ensures each spec runs with a new, fresh instance of config to prevent state leaks when particular specs modify config--so if you set it there, it'll ensure `no_color = true` is set for all specs in our suite by default w/o the need to monkey patch _all_ instances of config, and then I believe you can remove this `tap` block here.

myronmarston commented on features/formatters/configurable_colors.feature:
```diff
@@ -19,7 +19,7 @@ Feature: Configurable colors
       RSpec.configure do |config|
         config.failure_color = :magenta
         config.tty = true
```
This line should no longer be needed now that `force_color` is set.

myronmarston commented:
Well done, @CodingItWrong!  I appreciate your thoroughness on this. In spite of your name, you coded this right :).

CodingItWrong commented:
@myronmarston Thanks! I think I made every change you requested, except for the ones where I had a question or explanation. I thumbsed-up each one I made a change for for tracking. Please let me know any additional questions or changes!

myronmarston commented on lib/rspec/core/option_parser.rb:
```diff
@@ -140,8 +140,19 @@ def parser(options)
+          options[:color] = true
+        end
+
+        parser.on('--force-color', '--force-colour', 'Force the output to be in color, even if the output is not a TTY') do |_o|
```
I just realized that these options are wider than the default width, causing the help text to layout sub-optimally:

```
    -b, --backtrace                  Enable full backtrace.
    -c, --color, --colour            Deprecated: enable color in the output.
        --force-color, --force-colour
                                     Force the output to be in color, even if the output is not a TTY
        --no-color, --no-colour      Force the output to not be in color, even if the output is a TTY
    -p, --[no-]profile [COUNT]       Enable profiling of examples and list the slowest examples (default: 10).
    -w, --warnings                   Enable ruby warnings
```

Notice that the description for these is on the next line so it isn't smashed together with the options themselves.  If you set `parser.summary_width = 34` then it lays out nicely:

```
    -b, --backtrace                    Enable full backtrace.
    -c, --color, --colour              Deprecated: enable color in the output.
        --force-color, --force-colour  Force the output to be in color, even if the output is not a TTY
        --no-color, --no-colour        Force the output to not be in color, even if the output is a TTY
    -p, --[no-]profile [COUNT]         Enable profiling of examples and list the slowest examples (default: 10).
    -w, --warnings                     Enable ruby warnings
```

Mind doing that?

CodingItWrong replied:
Thanks for the pointer on how to implement!

myronmarston commented on lib/rspec/core/configuration.rb:
```diff
@@ -781,23 +783,62 @@ def full_backtrace=(true_or_false)
       end
 
+      # Returns the configuration option for forcing color, but should not
+      # be used to check if color is supported.
```
It would be nice if the docs for this (and the other methods like `color` and `no_color`) described what they do.  "Returns the configuration option for forcing color" is pretty obvious already, and I imagine a user checking the docs for this would want to understand why they might ever want to use this config option.  How about we change these to something like:

``` ruby
# Enables color output if the output is a TTY.  As of RSpec 3.6, this is the default
# behavior and this option is retained only for backwards compatibility.
#
# @depreacted #...
#
# @see force_color
# @see no_color
# @see color_enabled?
# @return [Boolean]
def color
  # ...
end

# Forces the output to be in color, even if the output is not a TTY.
#
# @note By default, RSpec will produce color if the output is a TTY, so this option
#   is really only needed if your output is a non-TTY (or sometimes a non-TTY).
# @see color
# @see no_color
# @see color_enabled?
# @return [Boolean]
def force_color
  # ...
end

# Prevents RSpec from producing color output, regardless of whether or not the
# output is a TTY.
#
# @see force_color
# @see color
# @see color_enabled?
# @return [Boolean]
def no_color
  # ...
end
```

CodingItWrong replied:
Do we want `force_color` and `no_color` to have `@see` annotations pointing to `color`? Since `color` is deprecated, maybe we don't need to draw attention to it.

myronmarston replied:
Good point.  Let's cut that annotation.

myronmarston commented on spec/rspec/core/configuration_spec.rb:
```diff
@@ -1229,7 +1229,174 @@ def metadata_hash(*args)
+    describe "color options" do
+      context "#force_color=false, #no_color=false" do
+        before { config.force_color = false }
+        before { config.no_color = false }
```
There's a bit of overhead for every `before` hook you defined, and I don't really see an advantage to making these separate hooks instead of combining them into one multi-line hook:

``` ruby
before do
  config.force_color = false
  config.no_color = false
end
```

Mind combining them (here and elsewhere in this PR)?

CodingItWrong replied:
Will do, just wasn't thinking.

CodingItWrong replied:
OK, I think this was the only file this problem occurred in.

myronmarston commented on spec/rspec/core/formatters/base_text_formatter_spec.rb:
```diff
@@ -268,6 +268,7 @@ def run_all_and_dump_failures
       RSpec.configure do |config|
+        config.no_color = false
         config.color = true
         config.tty = true
```
Can we just set `config.force_color = true` here?  In general, it would be nice to remove/replace any uses of the `tty` option since we are doing away with it and `force_color` is the new way to force color.

myronmarston commented:
Thanks for pushing the updates!  I re-reviewed (including leaving some responses to questions which github is hiding due to the lines in question having changed -- please check those).  This is getting close to ready to merge :).

CodingItWrong commented:
(Pulled away from finishing my changes, so don't feel the need to review anything until I finish batching it up!)

CodingItWrong commented:
@myronmarston OK, a few more questions for you, and I'm not sure if Appveyor was failing previously. I couldn't quite figure out the error in their UI.

myronmarston commented:
The build failures are happening due to your change in e8a31a6ebfdcc82ee1af13f5b0e8a345c6f8aa25 -- `abort` makes the process exit, and RSpec does not interfere with it, so that change means that now `rspec` could abort the run at any time if `color_enabled?` is checked and the `no_color` and `force_color` flags conflict -- which is obviously undesirable, and leads to the kind of confusing results you are seeing on appveyor, where the build exits with `1` with no clear indication as to why.  One hint is that we have over 2K specs but, depending on the seed, fewer than 1000 actually ran!

When I suggested above that we abort when `--force-color` and `--no-color` are both set, I was intending to have that limited to the CLI only.  An `abort` in the context of the option parser has much more limited scope, and does not risk being hit at nearly anytime like an `abort` in `config` does.

However, I can see why you put it in config (the two options could both be set via `config.no_color = true` and `config.force_color = true`, which creates the same problem) and I'm glad you put the `abort` there because it made me realize something :).  While we are limited in what kind of API we can provide in the CLI by the conventions and limitations of command line switches, we don't have to limit ourselves in the same way in our ruby config API.  It's clearly quite confusing to have `color`, `force_color` and `no_color`, and for all 3 to support 2 different values (true and false).  That creates way to many possible states for something that should be far simpler.

In practice, we really only care about supporting 3 states:

* `:enabled` -- color output is produced if output is a TTY
* `:forced` -- color output is always on, regardless of output
* `:disable` -- color output is always off, regardless of output

(Those names may not be the best -- if you have better suggestions, let me know).

We can support these with a single option off of `config`:

``` ruby
RSpec.configure do |c|
  c.color_mode = :forced # or :disabled or :enabled
end
```

That's 1 option with 3 possible values instead of 3 options with 2 possible values each.  Even better, this API provides no way to set an invalid state such as both `no_color` and `force_color`.  It's always nice when you model things in a way that does not even allow invalid states!

The existing deprecated `config.color` option can simply delegate to `color_mode`:

``` ruby
def color=(value)
  self.color_mode = value ? :enabled : :disabled
end
```

The CLI options will remain as `--force-color` and `--no-color`, which will set `color_mode` appropriately.

Once that is done, then the `abort` for `--force-color` and `--no-color` both being passed can happen in the option parser where it belongs, since that's the only place where that conflict can occur.

Does that all make sense?  Sorry about asking you to re-work this significantly...but this is the kinda realization that I don't think we could have made up front until someone actually dug into the complexity of implementing this.  I hope you're able to keep moving this PR forward, but if you hit a point where you lack the time or desire to finish this (since it's taking more time then you probably expected up front...) let us know.

CodingItWrong commented:
@myronmarston Thanks for the good design thoughts! Luckily I work at a place with a big focus on code quality, so thses kinds of back-and-forth and realizations aren't a new experience for me =] I need to buckle down on client work for a few days, but I'm definitely still excited to finish this PR, so I'll be back to it hopefully by the end of the week.

JonRowe commented:
I like the idea of having one colour config which has multiple values, how about `:automatic, :on, :off` (or substitute `:forced` for `:on`, but I'd like us to avoid `:disabled`)

myronmarston commented:
> but I'd like us to avoid `:disabled`

Why?

JonRowe commented:
It feels ableist to me

soulcutter commented:
I don't mind the alternatives, but I also don't think it's ableist to use the word disabled in an appropriate context. There's no negative connotation, and no reference to people.

I did a little research, and did not find the word 'disabled' to be on lists of ableist words to avoid http://www.autistichoya.com/p/ableist-words-and-terms-to-avoid.html

CodingItWrong commented:
OK, updated to have a single color config. A lot of the tests read more clearly with the new config, which is…a "good smell"? Is that a thing? ;-) I'm still in progress adding the `abort` call to the option parser, but wanted to push this up in the meantime. A few notes:

- I tried the enabled/disabled/forced language to start out, but aside from the ableist question, I'm not sure if the "enabled" value communicates its conditional nature as much as @JonRowe's "automatic", so I changed the options to automatic/disabled/forced.
- I didn't take @myronmarston's approach to the legacy "color" option, because that would have changed the behavior: it no longer would have respected the --tty flag. I think as-implemented it's a pretty simple fallback to checking "color" if "color_mode" isn't set to :forced or :disabled, but let me know if it's better to remove support for --tty to get the code simpler.
- If we do keep my approach of not removing the color flag entirely, the question from convo <https://github.com/rspec/rspec-core/pull/2321/files/8c0eb3b96fced5d27d5a72f79040935e9c996f6a..9af8f0cbde2f61f7ad6f7c406749f0439ce78d0c#r78248926> still remains, about preserving past behavior when color is set to false instead of left unset. I'm up for whichever approach you prefer. Thoughts?

CodingItWrong commented:
OK, the `abort` feature is pushed up now. Any thoughts on that Cucumber feature failing in JRuby?

myronmarston commented on lib/rspec/core/configuration.rb:
```diff
@@ -778,24 +780,58 @@ def full_backtrace=(true_or_false)
+          tty? || output_to_tty?(output)
+        else # color_mode == :automatic
+          output_to_tty?(output)
+        end
```
Not a merge blocker, but I think this would be slightly better as:

``` ruby
case color_mode
when :forced then true
when :disabled then false
else # :automatic
  output_to_tty?(output) || (color && tty?)
end
```

I like this better because it puts the new scheme (the color mode) front and center instead of putting a `elsif color` conditional between the color mode branches.  After all, the `elsif color` mode implies `color_mode == :automatic`, too, so it's a bit weird that that comment is only on the `else` branch, as you have it now.

myronmarston commented on lib/rspec/core/option_parser.rb:
```diff
@@ -140,8 +142,23 @@ def parser(options)
+        parser.on('-c', '--color', '--colour', 'Deprecated: enable color in the output.') do |_o|
+          options[:color] = true
+          options[:color_mode] = :automatic
+        end
```
I thought we had decided not to list `--color` in the output (see https://github.com/rspec/rspec-core/pull/2321#r77930510) since it's a deprecated option and we consider the `--help` output to be "valuable real estate" that we don't want taken up by deprecated options?

Any reason it can't be removed from the output?  (Of course, the option itself should still be honored, regardless of whether or not we display it in `--help`).

myronmarston replied:
Actually, I see now that you did remove it from the output by adding `-c` to the list of ~~`invalid_options`~~ `hidden_options`. Somehow I missed that.

Still, the fact that the code is here makes it easy to think that `--color` _will_ be listed in the output (as I initially thought it would), which makes this approach a bit confusing.  Instead, can you handle it similar to how [`--tty` is processed](https://github.com/rspec/rspec-core/blob/c3c90c278fda997e0eaf3bdac08d162175dc060b/lib/rspec/core/option_parser.rb#L21)?  It's much easier to understand that it is a "hidden" option that does not show up in the `--help` output.

CodingItWrong replied:
Update on this: if I move processing of the `--color` flag to where the `--tty` flag is processed, it seems as though it'll be difficult or impossible to retain the backwards-compatible behavior of respecting the order of `--color` and `--no-color` flags, so that the last one wins. (As a reminder, I did _not_ add last-one-wins for the new `--force-color` flag: only one of `--force-color` and `--no-color` can be specified. This is just about backwards compatibility.)

I think backwards compatibility has a higher priority than the clarity of where it is in the source code. Maybe there is another way to make it clearer that it won't be listed in the output. A comment? A change to the text "description" passed into `parser.on`?

myronmarston replied:
Woops, somehow I didn't get notified about this comment or just didn't notice the email or something.  Sorry about that!



> I think backwards compatibility has a higher priority than the clarity of where it is in the source code.

Agreed.

> Maybe there is another way to make it clearer that it won't be listed in the output. A comment? A change to the text "description" passed into parser.on?

Both of those would help.  The description is partly what threw me originally: since you took the time to update it to `Deprecated: ...` it made it seem like it would be printed.  Replacing it with `''` and a comment to explain it would address the issue well enough, I think.

myronmarston commented on lib/rspec/core/option_parser.rb:
```diff
@@ -140,8 +142,23 @@ def parser(options)
+
+        parser.on('--force-color', '--force-colour', 'Force the output to be in color, even if the output is not a TTY') do |_o|
+          if options[:color_mode] == :disabled
+            abort "Please only use one of --force-color and --no-color"
```
Can you put backticks around `--force-color` and `--no-color`?  We generally do that in messages like this.

myronmarston commented on spec/rspec/core/configuration_options_spec.rb:
```diff
@@ -422,7 +445,7 @@ def expect_parsing_to_fail_mentioning_source(source, options=[])
       with_env_vars 'SPEC_OPTS' => "--example 'foo bar'" do
         options = parse_options("--drb")
-        expect(options[:color]).to be_truthy
+        expect(options[:color_mode]).to eq(:automatic)
```
Since `--color` is deprecated, it would be nice to remove its use from examples that aren't specifically testing it.  This example is testing the various options files, and only used `--color` as an example of an option.  I'd like to see it changed to something else (perhaps `--order random` or similar?) so that when we do remove `--color` in the future we don't have to touch examples like this that are really about something else.

myronmarston commented on spec/rspec/core/configuration_spec.rb:
```diff
@@ -1229,7 +1229,167 @@ def metadata_hash(*args)
+          config.force :color_mode => true
+          config.color_mode = false
+          expect(config.color_mode).to be true
+        end
```
`true` is not a valid value for `color_mode`, so why does this set it to that value?  Can it set it to some kind ov valid value instead?

myronmarston commented on spec/rspec/core/configuration_spec.rb:
```diff
@@ -1229,7 +1229,167 @@ def metadata_hash(*args)
+      describe "#color_mode=" do
+        it "prefers incoming cli_args" do
+          config.output_stream = StringIO.new
+          allow(config.output_stream).to receive_messages(:tty? => true)
```
What does the output stream have to do with the rest of this example?  It does not seem to be used since it does not check `color_enabled?`...

CodingItWrong replied:
Ah yup, that was just a carryover from when this code was for the `color` option, but apparently the output stream didn't need to be stubbed there either.

myronmarston commented on spec/rspec/core/configuration_spec.rb:
```diff
@@ -1229,7 +1229,167 @@ def metadata_hash(*args)
+            allow(config.output_stream).to receive_messages(:tty? => true)
+
+            expect(config.color_enabled?).to be true
+            expect(config.color_enabled?(output)).to be true
```
In all of the examples for `color_mode` you have two paired expectations like this:

``` ruby
expect(config.color_enabled?).to be value
expect(config.color_enabled?(output)).to be value
```

...where `value` is always the same.  I'm realizing that as written, the implementation could completely ignore the argument, because the passed arg has the same `tty?` status as the configured `output_stream` (and in fact, it's the very same IO object!).

Since it's important that `color_enabled?` does in fact check `tty?` on the passed arg, and does not always check `output_stream`, it would be nice if the specs would cover this case.  Can you restructure the existing specs to cover this or add a new one to cover it?

myronmarston commented on spec/rspec/core/option_parser_spec.rb:
```diff
@@ -394,5 +394,24 @@ module RSpec::Core
+        parser = Parser.new(%w[--no-color --force-color])
+        expect(parser).to receive(:abort) do |msg|
+          expect(msg).to include('only use one of --force-color and --no-color')
+        end
```
I think this can be simplified to:

``` ruby
expect(parser).to receive(:abort).with(a_string_including("only use one of --force-color and --no-color"))
```

No need for the nested expectation inside the block :).

JonRowe commented on lib/rspec/core/configuration.rb:
```diff
@@ -778,24 +780,58 @@ def full_backtrace=(true_or_false)
         value_for(:color) { @color }
       end
 
+      # Sets the color mode to one of:
```
This doesn't set the color mode, it returns the current mode, if we keep this comment we should group the writer here for clarity.

CodingItWrong replied:
That makes sense to me. @myronmarston would there be any adverse impact to the organization of the docs by making this change?

myronmarston replied:
I'm all for improving the docs.  I'd just warn that YARD doesn't always render things how you think it will, and ultimately what we care the most about is how it looks when rendered -- so please run a local yard server (using `bundle exec yard server --reload`) while editing so you can take a look at the changes and ensure they render well.

JonRowe commented on lib/rspec/core/configuration.rb:
```diff
@@ -778,24 +780,58 @@ def full_backtrace=(true_or_false)
 
-      # Returns the configuration option for color, but should not
-      # be used to check if color is supported.
+      # Enables color output if the output is a TTY.  As of RSpec 3.6, this is
```
Same here this doesn't actually do the setting.

JonRowe commented on lib/rspec/core/configuration.rb:
```diff
@@ -778,24 +780,58 @@ def full_backtrace=(true_or_false)
+      #
+      # @deprecated No longer recommended because of complex behavior. Instead,
+      #   rely on the fact that TTYs will display color by default, or set
+      #   {:color_mode} to :forced to display color on a non-TTY output.
```
@myronmarston do we want to properly deprecate this?

myronmarston replied:
In general, I don't see a benefit to hard-deprecating something like this (e.g. with a warning).  A soft deprecation (via docs and a changelog entry) is sufficient for now, IMO, and then we can hard-deprecate it in 3.99.

In the 2.x releases we definitely got complaints about too many deprecation warnings.  In 3.x releases (where we've only been soft-deprecating things so far) I've not heard a single user complain that they wish we were issuing deprecation warnings.

CodingItWrong commented:
@myronmarston All code changes are in and ready for review!

Regarding `:forced_enabled`, that's definitely a clarity improvement. It does bug me aesthetically that it's two words, though—the other two flags are so much simpler!

This is making me come down strongly in favor of @JonRowe's suggestion of `:on`, `:automatic`, and `:off`. Those seem like the simplest words we could pick ("the color is either on, or it's off, or it's automatically set"), and they avoids the `:forced` ambiguity you mentioned. Not to mention it avoids confusion in the docs duplicating the name of the `Configuration#force` method (`it "forces force_color" do`).

myronmarston commented on spec/rspec/core/option_parser_spec.rb:
```diff
@@ -394,5 +394,22 @@ module RSpec::Core
+      it 'aborts if --force-color was previously set' do
+        parser = Parser.new(%w[--force-color --no-color])
+        expect(parser).to receive(:abort) do |msg|
+          expect(msg).to include('only use one of --force-color and --no-color')
```
Would be nice if this one was changed to use `with(a_string_including())` as well!

CodingItWrong replied:
Sounds good. I went ahead and changed the pre-existing `:abort`/`use --help` specs to use `a_string_including` too while I was in there, that way the whole file uses the same approach.

myronmarston commented:
> @myronmarston All code changes are in and ready for review!

The changes all look good, but I didn't see anything done to address [my comment](https://github.com/rspec/rspec-core/pull/2321#discussion_r79428183) about the confusing way `--color` is silenced from `--help`.  Our future selves are going to expect that all help text in the `parser.on` calls will be in the help output, and it's confusing for code later on to silence parts of it.  Thoughts?

> This is making me come down strongly in favor of @JonRowe's suggestion of `:on`, `:automatic`, and `:off`. Those seem like the simplest words we could pick ("the color is either on, or it's off, or it's automatically set"), and they avoids the `:forced` ambiguity you mentioned. Not to mention it avoids confusion in the docs duplicating the name of the `Configuration#force` method (`it "forces force_color" do`).

You make a good point.  I'm fine with `:on`, `:automatic` and `:off`.  Let's go with that.

Once those things are addressed, I think this will be ready to merge :).

Although -- at 38 commits and counting, it's kinda a lot of noise to introduce into the history for this PR.  Do you want to squash it down into a smaller number of logical commits?  Github also lets us squash into one commit when merging now so we can always do that if you want to leave as-is...but if you want it to go in as a handful of commits (instead of only 1 or 38+), feel free to rebase and combine things as you see fit.

CodingItWrong commented:
Whoops, I just lost track of the --color silencing comment—I'll get on that.

I think I'll rebase into a more "revisionist history" series of commits after all my remaining changes are approved—I find it easier to keep track for now without the rebasing 👍

CodingItWrong commented:
@myronmarston OK, all feedback addressed except I had a question about the approach to hiding `--color` from the help output: https://github.com/rspec/rspec-core/pull/2321#discussion_r79428183

Again, I think I'll wait for full approval of the changes before I rebase.

myronmarston commented:
> Again, I think I'll wait for full approval of the changes before I rebase.

:+1: I fully approve :).

CodingItWrong commented:
@myronmarston since i'm new to OSS contributions, just to make my intentions clear: i'll rebase this branch and then force-push it up. i _think_ github doesn't lose the comment history as part of that, but things could get confusing so i didn't want to do that without at least giving you the chance to stop me =]

myronmarston commented:
> 'll rebase this branch and then force-push it up. i think github doesn't lose the comment history as part of that, but things could get confusing so i didn't want to do that without at least giving you the chance to stop me =]

:+1:

JonRowe commented on spec/rspec/core/formatters/base_text_formatter_spec.rb:
```diff
@@ -268,8 +268,7 @@ def run_all_and_dump_failures
       RSpec.configure do |config|
-        config.color = true
-        config.tty = true
+        config.color_mode = :forced
```
This spec is the one Appveyor isn't happy with, shouldn't this be `:on`?

CodingItWrong commented:
@myronmarston @JonRowe 🚀

myronmarston commented:
Well done, @CodingItWrong!  Thanks for being so thorough.
